enforcer/src/main/java/org/wso2/micro/gateway/enforcer/security/AuthFilter.java

```diff
@@ -46,15 +49,65 @@ public boolean handleRequest(RequestContext requestContext) {
                     if (authenticate.isAuthenticated()) {
+                        updateClusterHeaderAndCheckEnv(requestContext, authenticate);
                         return true;
                     }
                 }
             }
         } catch (APISecurityException e) {
-            requestContext.getProperties().put("code", e.getErrorCode());
+            //TODO: (VirajSalaka) provide the error code properly based on exception (401, 403, 429 etc)
+            requestContext.getProperties().put("code", "401");
             requestContext.getProperties().put("error_code", e.getErrorCode());
             requestContext.getProperties().put("error_description", e.getMessage());
         }
         return false;
     }
+
+    /**
+     * Update the cluster header based on the keyType and authenticate the token against its respective endpoint
+     * environment.
+     * 
+     * @param requestContext request Context 
+     * @param authContext authentication context
+     * @throws APISecurityException if the environment and 
+     */
+    private void updateClusterHeaderAndCheckEnv(RequestContext requestContext, AuthenticationContext authContext)
+            throws APISecurityException {
+
+        String keyType = authContext.getKeyType();
+        if (StringUtils.isEmpty(authContext.getKeyType())) {
+            keyType = "PRODUCTION";
+        } 
+
+        // Header needs to be set only if the relevant cluster is available for the resource and the key type is
+        // matched.
+        if (requestContext.isClusterHeaderEnabled()) {
+            if (keyType.equalsIgnoreCase("PRODUCTION") &&
+                    !StringUtils.isEmpty(requestContext.getProdClusterHeader())) {
+                requestContext.addResponseHeaders(AdapterConstants.CLUSTER_HEADER,
+                        requestContext.getProdClusterHeader());
+            } else if (keyType.equalsIgnoreCase("SANDBOX") &&
+                    !StringUtils.isEmpty(requestContext.getSandClusterHeader())) {
+                requestContext.addResponseHeaders(AdapterConstants.CLUSTER_HEADER,
+                        requestContext.getSandClusterHeader());
+            } else {
+                throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                        "Invalid Token to access production/sandbox environment.");
+            }
+        } else {
+            // Even if the header flag is false, it is required to check if the relevant resource has a defined cluster
+            // based on environment. 
+            // If not it should provide authentication error.
+            // Always at least one of the cluster header values should be set.
+            if (keyType.equalsIgnoreCase("PRODUCTION")
+                    && StringUtils.isEmpty(requestContext.getProdClusterHeader())) {
+                throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
```

Comment:
Error message should be as same as the one presents in API Manager

Reply:
fixed via https://github.com/wso2/product-microgateway/pull/1525/commits/a4c0595ae454ec89caaf825f56270c2bae29e4c8